devito/symbolics/manipulation.py

```diff
@@ -319,12 +319,92 @@ def reuse_if_untouched(expr, args, evaluate=False):
         return expr.func(*args, evaluate=evaluate)
 
 
-def evalmin(a, b):
+def evalrel(func=min, input=None, assumptions=None):
     """
-    Simplify min(a, b) if possible.
+    Simplify evalmin if possible and return nested MIN/MAX expression. This function
+    takes advantage of python's built-in `min`, Sympy's `Min` and our in-house `MIN`
+    to reduce multiple candidates in relation's when possible.
+
+    Parameters
+    ----------
+    func : min or max, the operation to follow
+        Defaults to min
+    input : a list of the candidate symbols to be simplified
+        The candidates to be evaluated. Defaults to None.
+    assumptions : A list of assumptions formed as relationals between candidates,
+        assumed to be True. Defaults to None.
+
+    Examples
+    --------
+    Assuming no values are known for a, b, c, d but we know that d<=a and c>=b we can
+    safely drop `a` and `c` from the candidate list
+
+    >>> from devito import Symbol
+    >>> a = Symbol('a')
+    >>> b = Symbol('b')
+    >>> c = Symbol('c')
+    >>> d = Symbol('d')
+    >>> evalrel(max, [a, b, c, d], [[Le(d, a), Ge(c, b)]])
+    MAX(a, c)
     """
+    sfunc = (Min if func is min else Max)  # Choose sympy's Min/Max
+    rfunc = (MIN if func is min else MAX)  # Choose built-in MIN/MAX
+
+    if len(input) == 1:
+        return input[0]
+
+    mapper = {}
+
+    if assumptions:
+        temp = []
+        # Drop one in mutual pairs [(a, b) , (b, a)] == > [(a, b)]
+        for asm in assumptions:
+            if (asm.__class__ in (Ge, Gt)) and asm.rhs.is_Add and all(i.is_positive
+               for i in asm.rhs.args):
+                for j in asm.rhs.args:
+                    temp.append(Ge(asm.lhs, j))
+
+        for i in temp:
+            assumptions.append(i)
+
+        # Symmetry/ Antisymmetry
+        for asm in assumptions:
+            if set(asm.args).issubset(input):
+                if (asm.__class__ in (Ge, Gt)) and func is max:
+                    mapper.update({asm.args[1]: asm.args[0]})
+                elif (asm.__class__ in (Le, Lt)) and func is max:
+                    mapper.update({asm.args[0]: asm.args[1]})
+                elif (asm.__class__ in (Ge, Gt)) and func is min:
+                    mapper.update({asm.args[0]: asm.args[1]})
+                elif (asm.__class__ in (Le, Lt)) and func is min:
+                    mapper.update({asm.args[1]: asm.args[0]})
+
+    # Drop one in mutual pairs [(a, b) , (b, a)] == > [(a, b)]
+    temp = mapper.copy()
+    for k, v in mapper.items():
+        if mapper[k] is v and v in temp:
```

Comment:
`mapper[k] is v` ? what am I missing? `mapper[k]` **must** be `v` at this point no?